notebooks/peerscout/peerscout-get-editor-pubmed-paper-ids.ipynb

```diff
@@ -77,102 +75,124 @@
-   "outputs": [],
    "source": [
     "_sql = (\n",
-    "    'SELECT person_id FROM `{table}`'\n",
-    ").format(table=editor_parsed_pubmed_ids_table_name)\n",
+    "    '''\n",
+    "    SELECT\n",
+    "        person_id,\n",
+    "        pubmed_url,\n",
+    "        provenance.imported_timestamp as imported_timestamp\n",
+    "    FROM `{table}`\n",
+    "    ''').format(table=editor_parsed_pubmed_ids_table_name)\n",
     "\n",
     "try:\n",
-    "    existing_editor_ids_df = read_big_query(_sql)\n",
+    "    existing_editor_ids_and_pubmed_url_df = read_big_query(_sql)\n",
     "except Exception as e:  # pylint: disable=broad-except\n",
     "    if not is_bq_not_found_exception(e):\n",
     "        raise\n",
     "    print('table not found: %s' % editor_parsed_pubmed_ids_table_name)\n",
-    "    existing_editor_ids_df = pd.DataFrame(columns=['person_id'])\n",
-    "print(len(existing_editor_ids_df))\n",
-    "existing_editor_ids_df.head()"
-   ]
+    "    existing_editor_ids_and_pubmed_url_df = pd.DataFrame(columns=['person_id','pubmed_url', 'imported_timestamp'])\n",
+    "print(len(existing_editor_ids_and_pubmed_url_df))\n",
+    "existing_editor_ids_and_pubmed_url_df.head(3)"
+   ],
+   "outputs": [],
+   "metadata": {}
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
-   "outputs": [],
    "source": [
-    "existing_editor_ids_set = set(existing_editor_ids_df['person_id'])\n",
+    "existing_editor_ids_set = set(existing_editor_ids_and_pubmed_url_df['person_id'])\n",
     "len(existing_editor_ids_set)"
-   ]
+   ],
+   "outputs": [],
+   "metadata": {}
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "source": [
+    "existing_editor_pubmed_url_set = set(existing_editor_ids_and_pubmed_url_df['pubmed_url'])\n",
+    "len(existing_editor_pubmed_url_set)"
+   ],
    "outputs": [],
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
    "source": [
     "editor_parsed_pubmed_links_df = read_big_query(\n",
     "    'SELECT * FROM `{table_name}`\\nWHERE parsed_search_term IS NOT NULL'.format(\n",
     "        table_name=editor_parsed_pubmed_links_table_name\n",
     "    )\n",
     ")\n",
     "print(len(editor_parsed_pubmed_links_df))\n",
-    "editor_parsed_pubmed_links_df.head()"
-   ]
+    "editor_parsed_pubmed_links_df.head(3)"
+   ],
+   "outputs": [],
+   "metadata": {}
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
-   "outputs": [],
    "source": [
     "remaining_editor_parsed_pubmed_links_df = editor_parsed_pubmed_links_df[\n",
     "    ~editor_parsed_pubmed_links_df['person_id'].isin(existing_editor_ids_set)\n",
     "]\n",
+    "\n",
+    "remaining_editor_parsed_pubmed_links_df = remaining_editor_parsed_pubmed_links_df.append(\n",
+    "    editor_parsed_pubmed_links_df[\n",
+    "        ~editor_parsed_pubmed_links_df['pubmed_url'].isin(existing_editor_pubmed_url_set)\n",
```

Comment:
In practise this will probably work most of the times, but it isn't 100% correct I believe.
For example, what if the editorial team notices that they had the URLs the wrong way around for two editors - then the URL would be in the set but not for the correct editor.

Maybe it will be worth creating a function a PeerScout related Python package in order to add tests?

For this particular case, one could have a set with tuple of `person_id` and `pubmed_id` and then check whether the set contains that. But that wouldn't be too helpful when adding imported timestamp.

The solution that would come to mind, is to use `merge` to left-join the two DataFrames. Then you can do simple Pandas operations, e.g. something like that:

```python
merged_editor_parsed_pubmed_links_df = editor_parsed_pubmed_links_df.merge(
    existing_editor_ids_and_pubmed_url_df,
    how='left',
    on='person_id',
    suffixes=('', '_existing')
)

remaining_editor_parsed_pubmed_links_df = merged_editor_parsed_pubmed_links_df[
    (merged_editor_parsed_pubmed_links_df['pubmed_url'] != merged_editor_parsed_pubmed_links_df['pubmed_url_existing'])
    | (merged_editor_parsed_pubmed_links_df['imported_timestamp'] < ...)
].drop(columns=[...])
```

Reply:
If the URLs are in the wrong way around then the editorial team will realise and correct them, after that when the job run again it will be corrected in the dataframe too. In current solution it will cause duplication but because of we are planing to create the view after that step I was planing to keep just latest record for one editor. If I understand correctly it already will solve the problem? Between merging or appending will not be any difference in the case we use view step?

Reply:
Apologies, I may not have explained it well. The merging is not related to merging in the output table. It will be appended either way. Here the merging is just a useful tool to compare between the new url and url from the output table.

Checking the existence of the person id in the set is fine. But checking the URL doesn't seem correct, because it is important whether the URL for that particular editor with the same person id has changed.

Perhaps it is easier to explain in person?

Reply:
> it is important whether the URL for that particular editor with the same person id has changed.

Ah I see now :) OK thank you! I will look at it in this point of view.